### ISSUE TITLE
Cheat Engine 7.1 Slow Performance

### ISSUE THREAD

**Issue by Karlos007** [cheat-engine/cheat-engine#1221]
Opening a .CT of 3,82 MB takes around 37 sec. with this 7.1 version (12 sec. to close it)

While using 7.0 or 6.8.3 takes around 3 sec to open.

Using default settings, Windows 10, small tables work fine.

btw The copy paste bug seems to be fixed from 7.0 to 7.1.

**Comment by cheat-engine** [comment]
Do you open it while the process has been attached ?

Does the table contain non hexadecimal offsets ?

**Comment by Karlos007** [comment]
Hi, either attached or not takes the same time

And I'm not sure but I think all offsets are hex:
    <DropDownListLink>EXP</DropDownListLink>
    <Address>symbolasdf+16B8</Address>
I have three registerSymbol and hundreds of references to those symbols.

**Comment by mgrinzPlayer** [comment]
Can we download CT file from FRF?

**Comment by Karlos007** [comment]
> Can we download CT file from FRF? You can also use winrar or 7zip to create .zip file, you can drag and drop here in github comments.

Every big .CT file is extremely slow (10 MB takes a few minutes).
Some files are mine while others are from the forums.
[ct.zip](https://github.com/cheat-engine/cheat-engine/files/4531416/ct.zip)

**Comment by cheat-engine** [comment]
it's the description hashlist. guess I need to update that differently

**Comment by Karlos007** [comment]
> it's the description hashlist. guess I need to update that differently

Thanks for the fast answer, if you need anything else tell me.

**Comment by mgrinzPlayer** [comment]
"golden sun 2.ct" has 29833 memory records with address field. 

So I used this script in both CE7.0 and CE7.1:

```Lua
start = getTickCount()

count = 30000
for i=1,count do
  local mr = AddressList.createMemoryRecord()
  if i % 100 == 0 then processMessages() MainForm.Caption = i/count*100  end
end

print('time:', (getTickCount() - start) // 10 / 100 )
print( getCEVersion() )
```
```
time: 58.43 
7.0
```
```
time: 180.1 
7.1
```
Creating dummy memory records in CE7.0 takes less time.




Also saving and loading CT file with those dummy memory records reflects your observations.
```Lua
start = getTickCount()

count = 30000
for i=1,count do
  local mr = AddressList.createMemoryRecord()
  if i % 100 == 0 then processMessages() MainForm.Caption = i/count*100  end
end

print('time:', (getTickCount() - start) // 10 / 100 )
print( getCEVersion() )
```
```
time: 21.46 
7.0 
```
```
time: 193.78 
7.1 
```

**Comment by Karlos007** [comment]
Pretty interesting script. My cpu is i7-6700 btw.
But yeah seems some kind of issue with this new version.

**Comment by mgrinzPlayer** [comment]
Also tested it with CE6.8.3. No big difference between 6.8.3 and 7.0
Ahh, I see there's descriptionhashlist.Clear; executed every time memory record is destroyed or description of any existing MR has set/change.

**Comment by cheat-engine** [comment]
Fixed it, but try not to use duplicate memoryrecord description names if you make use of the (description) as you won't know which one will be picked. And if you do have two, and delete one of them , all will be gone until it has been reset

**Comment by mgrinzPlayer** [comment]
```Lua
file = [[pathto\30000 memrecs.CT]]

start = getTickCount()
loadTable(file)
print('loading time:', (getTickCount() - start) // 10 / 100,'s')

start = getTickCount()
getTableEntry('master').destroy()
print('destroy time:', (getTickCount() - start) // 10 / 100,'s')
print( getCEVersion() )
```


Tested with new patch:
- creation: 98.25, CE7.1 (versus 58.43, CE7.0)
- loading: 27.35, CE7.1  (versus 21.46, CE7.0)
- destroying: 40.42, CE7.1  (versus 5.50, CE7.0)

**Comment by mgrinzPlayer** [comment]
latest patch, two consecutive runs:

create time: 103.89 s 
loading time: 30.07 s 
destroy time: 46.51 s 

create time: 103.87 s 
loading time: 31.78 s 
destroy time: 47.71 s

**Comment by cheat-engine** [comment]
instead of doing a master destroy, do a AddressList.clear()

**Comment by mgrinzPlayer** [comment]
There's no _clear_ Lua method for AddressList.

Also, if you opened 30000MR.CT and then open other small.CT without merge, destroying also takes the same amount of time.

**Comment by cheat-engine** [comment]
> Also, if you opened 30000MR.CT and then open other small.CT without merge, destroying also takes the same amount of time.
I can't reproduce. it's down to 5 seconds for me when loading a small .CT afterwards. Also clicking the delete button above the addresslist is faster

**Comment by mgrinzPlayer** [comment]
Did you try it with master MR with 29999 children MR?

**Comment by cheat-engine** [comment]
yes, 5 seconds after clicking the erase button. Faster than loading the table

**Comment by mgrinzPlayer** [comment]
Clean CE build (I also removed lib folder to be sure).

Don't know, maybe Lazarus thing... (I'm using Laz208 and FPC 3.2.0-beta-r44046 from fixes_3_2 branch)

Can you attach zipped binary here?

**Comment by cheat-engine** [comment]
Check that you have this line: https://github.com/cheat-engine/cheat-engine/blob/19d92d87b2ae1e11feb736c60dd9b0ecf05d4382/Cheat%20Engine/addresslist.pas#L273

[cheatengine-x86_64.zip](https://github.com/cheat-engine/cheat-engine/files/4532004/cheatengine-x86_64.zip)

**Comment by mgrinzPlayer** [comment]
Yes, I have this line.
Clean build - erase button takes 40s
Your build 5s

Don't know. Maybe not all changes are pushed on github? Or Laz208 has broken TStringHashList...

**Comment by cheat-engine** [comment]
Do you build in release mode without a debugger attached ? Just having a debugger attached can slow down CE quite a bit

**Comment by mgrinzPlayer** [comment]
Laz206 OK
Laz208 not ok
Laz209 ( lazarus/branches/fixes_2_0 ) not ok

All Lazaruses were compiled with FPC 3.2.0-beta-r44046 from fixes_3_2 with the very same flags.

**Comment by mgrinzPlayer** [comment]
Double checked:
Laz206 OK
Laz208 no OK

**Comment by mgrinzPlayer** [comment]
Seems to be lazarus treeview.inc and recent scrollbars problem - they were fixing Cocoa drawing (upside down) glitch 34334. They moved calling UpdateScrollbars from TCustomTreeView.DoPaint to TCustomTreeView: Expand, Collapse, MoveSelection, Resize, DoMouseWheel, ScrollView
and TTreeNode: Destroy, InternalAddObject

If Lazarus 2.0.8 and above, for win builds we have to revert revision 62457 (merged revisions 61288, 61326, 61394, 61437)
